[staging] setup.sh: Support XDG_DATA_DIRS (bash completion in nix-shell)

roberth opened the pull request:
###### Motivation for this change

Add bash completion support for packages in nix-shells.

Closes #26031

XDG_DATA_DIRS is to /share as PATH is to /bin.

It was defined as part of the XDG basedir specification.
https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

While it originated from the X Desktop Group, it is not limited to
the X11 ecosystem, as evidenced by its use in bash-completion.

TODO
 - [x] one more test locally
 - [x] check buildInputs vs nativeBuildInputs behavior (not expecting any surprises though)
   - ~either of which works with bash completion because strictDeps is disabled~
   - [x] buildInputs should not be loaded because those will be for the wrong system
 - [ ] test a significant portion of Nixpkgs builds
   - [x] gnome3.eog
   - [x] konsole

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

roberth committed:
setup.sh: Support XDG_DATA_DIRS

XDG_DATA_DIRS is to /share as PATH is to /bin.

It was defined as part of the XDG basedir specification.
https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

While it originated from the X Desktop Group, it is not limited to
the X11 ecosystem, as evidenced by its use in bash-completion.

The removal of ` && -d "$pkg/bin"` is ok, because this optimization is
already performed by `addToSearchPath`.

matthewbauer commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -483,8 +483,9 @@ activatePackage() {
     # the transition, we do include everything in thatcase.
     #
     # TODO(@Ericson2314): Don't special-case native compilation
-    if [[ ( -z "${strictDeps-}" ||  "$hostOffset" -le -1 ) && -d "$pkg/bin" ]]; then
```
we may still want `-d $pkg/bin`. IIRC we get tons more in PATH without it, which can subtly break things that don't expect so long of a path

matthewbauer replied:
and maybe we should do the same for /share

roberth replied:
Both are taken care of by `addToSearchPath`.

> The removal of ` && -d "$pkg/bin"` is ok, because this optimization is
already performed by `addToSearchPath`.

matthewbauer commented:
While I think this is a good idea, I just wanted to note something similar was previously discussed in https://github.com/NixOS/nix/pull/2443 and @edolstra had some objections. It may be less controversial in Nixpkgs though since we rely on XDG_DATA_DIRS in a lot of places already.

matthewbauer commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -602,9 +603,11 @@ fi
 
 PATH="${_PATH-}${_PATH:+${PATH:+:}}$PATH"
 HOST_PATH="${_HOST_PATH-}${_HOST_PATH:+${HOST_PATH:+:}}$HOST_PATH"
+XDG_DATA_DIRS="${_XDG_DATA_DIRS-}${_XDG_DATA_DIRS:+${XDG_DATA_DIRS:+:}}${XDG_DATA_DIRS-}"
```
I spent a couple minutes looking through Nixpkgs to see if this could interfere with anything. I don't think it will, but setting something like this can always have unintended consequences. The main concern would be if native XDG_DATA_DIRS gets mixed in with target XDG_DATA_DIRS in cross-compilation. For things like bash completion and man pages, it's not such a big deal, but XDG_DATA_DIRS is also used to locate qt5 plugins.

matthewbauer replied:
Since it's not exported, this might not be as much of a problem actually.

roberth replied:
It is implicitly exported when `XDG_DATA_DIRS` already exists as an environment variable (and `--pure` isn't used).

Always exporting it seems more consistent. If mixing with the `nix-shell` host environment is problematic, that can be avoided with `nix-shell --pure`.

In cross compilation without `strictDeps`, mixing will occur. Perhaps `XDG_DATA_DIRS` should follow "strictDeps" behavior regardless. That will require development shells to put tools in `nativeBuildInputs` instead of `buildInputs` for bash completion to work. `nix-shell -p` gets that wrong, so it won't work there yet, but can be fixed.

roberth committed:
setup.sh: Only load XDG_DATA_DIRS for executable inputs

This avoids the scenario where strictDeps is off and cross-compiled
XDG_DATA_DIRS content is brought into the environment.

While probably harmless for data like manpages and completion scripts,
this would cause issues when XDG_DATA_DIRS is used to find executables
or plugins. The Qt framework is known to behave like this and might
have run into incompatibilities.

roberth committed:
setup.sh: export XDG_DATA_DIRS for consistency

By exporting it, we always make the new directories available
to subprocesses, regardless of whether the environment
variable existed before `nix-shell` was invoked.

roberth commented:
Ok, `XDG_DATA_DIRS` now follows strictDeps-like behavior. No mixing with cross-compilation outputs will occur.
(If some hypothetical build saves `XDG_DATA_DIRS`, that's equivalent to saving the builder's `PATH`; just one out of a thousand potential bad build ideas)

Also `XDG_DATA_DIRS` is now always exported as one will expect it to. Previous behavior was potentially confusing.

Ericson2314 reviewed (approved):
This looks good to me!

worldofpeace commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -602,9 +606,11 @@ fi
 
 PATH="${_PATH-}${_PATH:+${PATH:+:}}$PATH"
 HOST_PATH="${_HOST_PATH-}${_HOST_PATH:+${HOST_PATH:+:}}$HOST_PATH"
+export XDG_DATA_DIRS="${_XDG_DATA_DIRS-}${_XDG_DATA_DIRS:+${XDG_DATA_DIRS:+:}}${XDG_DATA_DIRS-}"
```
Is this going to affect wrapGAppsHook or wrapQtAppsHook in any way?

roberth replied:
It shouldn't affect those.
This variable is set in the build environment, not in the wrapper scripts.

worldofpeace replied:
cc @NixOS/freedesktop

worldofpeace replied:
Oh, so being set in the build environment, but during fixup it shouldn't be?

roberth replied:
Well it's still set during `fixupPhase` but it doesn't affect the wrapper writing functions. Although, it can affect call sites if they are like `wrapProgram --set XDG_DATA_DIRS $XDG_DATA_DIRS`, so expanding `$XDG_DATA_DIRS` before invoking `wrapProgram`, but I don't think we have that?

jtojnar replied:
Yeah, `wrapGAppsHook` uses different environment variables for populating `XDG_DATA_DIRS` for wrapper. So unless one of those variables is populated from `XDG_DATA_DIRS` (do not see any such thing when `grep`ing `sh` files for `XDG_DATA_DIRS` so it would have to be indirectly through some program).

The only possible issue that comes to mind is that it further distances the build environment from the runtime one, possibly obscuring some missing dependencies until runtime (i.e. `(install)CheckPhase` passes but program will crash at runtime). But perhaps the difference is huge enough already so this might not be so much worse in proportion.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/setup-bash-completion-when-running-under-nix-shell-on-macosx/5216/6

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/shell-nix-load-completions/8803/15

jonringer commented:
My understanding is that only nativeBuildInputs have shell completion from this, however, most examples use `buildInputs`.

nix manual: https://nixos.org/manual/nix/unstable/command-ref/nix-shell.html
wiki: https://nixos.wiki/wiki/Development_environment_with_nix-shell
wiki / cachix: https://nixos.wiki/wiki/Caching_nix_shell_build_inputs

I don't think this is a blocker, but we should probably update the documentation examples so that people get expected behavior

roberth commented:
> My understanding is that only nativeBuildInputs have shell completion from this, however, most examples use `buildInputs`.

That is correct. `buildInputs` shouldn't be loaded on `PATH` either. That is a legacy behavior.

The idea is that `XDG_DATA_DIRS` supports apps that run in the "current" context, `build` in our cross compilation terminology. `buildInputs` can, in general, only run in the `host` context. Loading these by default will cause issues with cross compilation.

This change does affect all derivations after all, for better or for worse. I'm hopeful but this needs testing.

Speaking of testing, on non-cross we should probably change or unset `XDG_DATA_DIRS` during `checkPhase` and `installCheckPhase`.

> I don't think this is a blocker, but we should probably update the documentation examples so that people get expected behavior

:100:. Should have been done with the introduction of cross compilation.

Ericson2314 commented:
@jonringer Yeah, we've never found a sane way to flip the `strictDeps` switch without breaking the world. This PR to me says "fine, but *new* functionality can be effectively be strictDeps-style". I think this is a rather clever way to "nudge not shove" the ecosystem in the right direction --- nothing will be broken, per se, but the incentive to switch will compound over time.

I like it a lot!

jonringer reviewed (approved):
LGTM

zimbatm commented:
One thing I didn't understand is; what kind of black magic auto-loads the completion? Is it part of bash itself to look for those files?

Incidentally, I recently worked on unifying the bash completions outputs: https://github.com/NixOS/nixpkgs/pull/103421

roberth commented:
@zimbatm 

> One thing I didn't understand is; what kind of black magic auto-loads the completion? Is it part of bash itself to look for those files?

No black magic, just a shell script in the bash-completion package.

https://github.com/scop/bash-completion/blob/f7475cf2c0d60f690a0bd758e3fa84976017b99a/bash_completion#L2185-L2187

> Incidentally, I recently worked on unifying the bash completions outputs: #103421

That's great!

roberth commented:
- Fixed up the first commit to unset the "private" `_XDG_DATA_DIRS` (note the underscore)
 - Tested build and execution of
   - `gnome3.eog`
   - `konsole` (KDE)

roberth commented:
> Speaking of testing, on non-cross we should probably change or unset XDG_DATA_DIRS during checkPhase and installCheckPhase.

We don't do this for `PATH` either, so let's keep it simple.
Another reason not to automagically change it for the check phases is that we want the package author to figure out that the data dependency exists, so that they are aware that `XDG_DATA_DIRS` must be set when the package is used.

jonringer commented:
@GrahamcOfBorg eval

Ericson2314 commented:
> wiki: https://nixos.wiki/wiki/Development_environment_with_nix-shell
> wiki / cachix: https://nixos.wiki/wiki/Caching_nix_shell_build_inputs

BTW I just updated those to use `nativeBuildInputs` where appropriate. Perhaps is a hoop users won't choose to jump through, and that's fine, but at least those docs won't be misleading them.

roberth commented:
Any objections to testing this on staging?
